include/uam/workers/uam_worker_base.h

```diff
@@ -0,0 +1,563 @@
+      ROS_ERROR_STREAM("Failed to decode message");
+      return false;
+    }
+    if (callback_)
+      callback_(*reply, wall_time);
+    else
+      ROS_WARN_STREAM(
+        "No handler for " << HeaderMSB << HeaderLSB << SubHeaderMSB << SubHeaderLSB << " and it is "
+                          << (int)(callback_ == nullptr));
+
+    return true;
+  }
+
+protected:
+  /**
+   * @brief Default C'tor
+   *
+   * This initialize
+   *
+   */
+  explicit WorkerBase(
+    const uint32_t header_offset = 0,
+    const uint32_t footer_offset = sizeof(protocol::CommandReplyHeader)) :
+    request_(Request { protocol::CommandRequestHeader { protocol::STX_ID,  // NOLINT
+                                                        sizeof(Request),
+                                                        { HeaderMSB, HeaderLSB },
+                                                        { SubHeaderMSB, SubHeaderLSB } },
+                       protocol::CommandFooter { 0, protocol::ETX_ID } }),
+    header_visitor_(header_offset),
+    footer_visitor_(footer_offset)
+  {
+    request_.footer.crc = calculateCrc(request_);
+    encoded_request_ = encodeCommand(request_);
+  }
+
+  /**
+   * @bief Convert value to Hexadecimal in string form
+   *
+   * @param[in] value - Value to be converted
+   * @param[in] size - The number of bytes of the value
+   * @return The hexadecimal value in string form
+   */
+  template <typename T>
+  static std::string toHexString(const T& value, const size_t& size = sizeof(T))
+  {
+    std::stringstream stream;
+    stream << std::uppercase << std::setfill('0') << std::setw(size) << std::hex << value;
+    return stream.str();
+  }
+
+  /**
+   * @brief Method to encode the command type Request
+   *
+   * @param[in] request - Request to encode.
+   * @return encoded command
+   */
+  static std::string encodeCommand(const Request& request)
+  {
+    std::string encoded_request;
+    encoded_request.reserve(sizeof(Request));
+    encoded_request += request.header.stx;
+    encoded_request += toHexString(request.header.cmd_size);
+    encoded_request += request.header.header[0];
+    encoded_request += request.header.header[1];
+    encoded_request += request.header.sub_header[0];
+    encoded_request += request.header.sub_header[1];
+    encoded_request += toHexString(request.footer.crc);
+    encoded_request += request.footer.etx;
+    return encoded_request;
+  }
+
+  /**
+   * @brief Calculate Reply CRC using raw buffer
+   *
+   * @param[in] buffer - Input buffer
+   * @return[out] crc value
+   */
+  uint16_t calculateReplyCrc(const std::string* buffer) const
+  {
+    // validate crc with expected crc:
+    const auto buffer_size = buffer->size();
+    const auto crc_buffer_size =
+      buffer_size - (sizeof(protocol::CommandReplyHeader::stx) + sizeof(protocol::CommandFooter));
+    return calculateCrc(
+      buffer->substr(sizeof(protocol::CommandReplyHeader::stx), buffer_size - sizeof(protocol::CommandFooter)).data(),
```

Comment:
Instead of copying the data into a substring, could you just index into the string buffer?
```suggestion
      buffer->data() + sizeof(protocol::CommandReplyHeader::stx),
```